### PR TITLE
fix: make TS validators accept null values

### PR DESCRIPTION
**Pull request by haijian-vaadin**
## Description
Fixes the bug that TypeScript form validators throw exceptions when validating null values. This could happen when loading an empty value from the backend. Related to https://github.com/vaadin/fusion/issues/44

Fixes # (issue)

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: flow-client/src/main/frontend/form/Validators.ts [hunk 1]
```diff
@@ -97,8 +97,8 @@ export class Email extends AbstractValidator<string> {
   constructor(attrs?: ValidatorAttributes) {
     super({ message: 'must be a well-formed email address', ...attrs });
   }
-  validate(value: string | undefined) {
-    return value === undefined || isEmail(value);
+  validate(value: string | null | undefined) {
+    return !value || isEmail(value);
```
Comment (Lodin) [inline flow-client/src/main/frontend/form/Validators.ts:101]: It also rejects the empty string. Is it ok? If not, we can replace it with the loose null check:
```suggestion
    return value != null || isEmail(value);
```

File: flow-client/src/main/frontend/form/Validators.ts [hunk 2]
```diff
@@ -71,15 +71,15 @@ abstract class NumberValidator<T> extends AbstractValidator<T> {
...
   optional: boolean;
   constructor(optional: boolean, attrs?: ValidatorAttributes) {
     super({ message: 'must be a number', ...attrs });
     this.optional = optional;
   }
 
-  validate(value: number | undefined) {
-    return (this.optional && value === undefined) || super.validate(value);
+  validate(value: number | null | undefined) {
+    return (this.optional && (value === null || value === undefined)) || super.validate(value);
```
Comment (Lodin) [inline flow-client/src/main/frontend/form/Validators.ts:82]: We can use the loose null check here:
```suggestion
    return (this.optional && value == null) || super.validate(value);
```

File: flow-client/src/main/frontend/form/Validators.ts (hunk 1 above)
Comment (haijian-vaadin) [reply flow-client/src/main/frontend/form/Validators.ts:101]: I also thought about this. I think the Email validator should accept the empty string, which could happen when the user first type sth, then remove the value. 
This kind of required check should be done in the Required/NotNull validator, not this Email validator.

File: flow-client/src/main/frontend/form/Validators.ts (hunk 2 above)
Comment (haijian-vaadin) [reply flow-client/src/main/frontend/form/Validators.ts:82]: cool, didn't know about this trick, thanks

File: flow-client/src/main/frontend/form/Validators.ts (hunk 1 above)
Comment (Lodin) [reply flow-client/src/main/frontend/form/Validators.ts:101]: Do I understand correctly that results of `!value` and `value != null` are the same then?

Comment (haijian-vaadin) [reply flow-client/src/main/frontend/form/Validators.ts:101]: No. The difference is `''`. `''` passes the 'Email' validation with `!value`, fails with `value !=null`. As indicated in the changes of another file.
```
const validator = new Email();
assert.isTrue(validator.validate(''));
```

**Review by platosha — APPROVED** [review]

**Pull request merged**